src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -269,16 +268,115 @@ pub mod diagnostics {
+                        // `ChildAccessForbidden(Reserved | Active)` can never exist.
+                        (_, Active) | (_, Reserved { .. }) => unreachable!("cannot cause an error"),
+                        // No transition has `Reserved` as its `.to` unless it's a noop.
+                        (Reserved { .. }, _) => unreachable!("noop transition"),
+                        // A pointer being `Disabled` is a strictly stronger source of
+                        // errors than it being `Frozen`. If we try to access a `Disabled`,
+                        // then where it became `Frozen` (or `Active`) is the least of our concerns for now.
+                        (Active | Frozen, Disabled) => false,
+                        // Implies that the permission became Disabled and was _later_
+                        // found Frozen, which is impossible because permissions only
+                        // evolve in the order `Reserved -> Active -> Frozen -> Disabled`.
+                        (Disabled, Frozen) => unreachable!("violates monotony of permissions"),
+                    }
+                }
+                TransitionError::ProtectedTransition(forbidden) => {
+                    assert!(!forbidden.is_noop());
+                    // Show how we got to the starting point of the forbidden transition,
+                    // but ignore what came before.
+                    // This eliminates transitions like `Reserved -> Active`
+                    // when the error is a `Frozen -> Disabled`.
+                    match (self.to, forbidden.from, forbidden.to) {
+                        (Active, Active, _) => true,
+                        (Frozen, Frozen, _) => true,
+                        // No transition has `Reserved` as its `.to` unless it's a noop.
+                        (Reserved { .. }, _, _) => unreachable!("noop transition"),
+                        // `Reserved -> Frozen` does not trigger protectors.
+                        (_, Reserved { .. }, Frozen) => unreachable!("cannot cause an error"),
+                        // No transition has `Disabled` as its `.from` unless it's a noop.
+                        (_, Disabled, _) => unreachable!("noop transition"),
+                        // These imply that the permission became `Disabled`
+                        // (resp. `Frozen`, `Active`) and was later found `Reserved | Active | Frozen`
+                        // (resp. `Reserved | Active`, `Reserved`), which is impossible because
+                        // permissions only evolve in the order `Reserved -> Active -> Frozen -> Disabled`.
+                        (Disabled, Reserved { .. } | Active | Frozen, _)
+                        | (Frozen, Reserved { .. } | Active, _)
+                        | (Active, Reserved { .. }, _) =>
+                            unreachable!("violates monotony of permissions"),
+                        // If the error is a transition `Frozen -> Disabled`, then we don't really
```

Comment:
This says `Frozen -> Disabled` but the code actually checks `Frozen -> _`. Or am I missing something?

Reply:
Would you prefer an explicit
```rs
| (Active, Frozen, Disabled) => false,
| (Active, Frozen, _) => unreachable!(),
```

or just a comment that the `_` is guaranteed to be `Disabled` ?

Reply:
Related: is it better if there is complete coverage of the `unreachable!()` cases, or is it fine if some of them would actually pass if accidentally reachable ?

Reply:
I'd prefer the former, given that the function already has tons of panics asserting the transition makes sense.

> Related: is it better if there is complete coverage of the unreachable!() cases, or is it fine if some of them would actually pass if accidentally reachable ?


Depends on how complex the code becomes if it wants to cover everything unreachable. E.g. I wouldn't want to `match` on fields that are currently entirely ignored just to add some assertions.